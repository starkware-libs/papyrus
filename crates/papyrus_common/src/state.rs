use indexmap::indexmap;
use rand::RngCore;
use serde::{Deserialize, Serialize};
use starknet_api::core::{ClassHash, CompiledClassHash, ContractAddress, Nonce};
use starknet_api::state::{StorageKey, ThinStateDiff};
use starknet_types_core::felt::Felt;

/// A storage entry in a contract.
#[derive(Debug, Default, Clone, Eq, PartialEq, Hash, Deserialize, Serialize, PartialOrd, Ord)]
pub struct StorageEntry {
    pub key: StorageKey,
    pub value: Felt,
}

/// A deployed contract in StarkNet.
#[derive(Debug, Default, Clone, Eq, PartialEq, Hash, Deserialize, Serialize, PartialOrd, Ord)]
pub struct DeployedContract {
    pub address: ContractAddress,
    pub class_hash: ClassHash,
}

/// A mapping from class hash to the compiled class hash.
#[derive(Debug, Clone, Default, Eq, PartialEq, Deserialize, Serialize)]
pub struct DeclaredClassHashEntry {
    pub class_hash: ClassHash,
    pub compiled_class_hash: CompiledClassHash,
}

#[derive(Debug, Clone, Default, Eq, PartialEq, Deserialize, Serialize)]
pub struct ReplacedClass {
    pub address: ContractAddress,
    pub class_hash: ClassHash,
}

// TODO: move to used crate
pub fn create_random_state_diff(rng: &mut impl RngCore) -> ThinStateDiff {
    let contract0 = ContractAddress::from(rng.next_u64());
    let contract1 = ContractAddress::from(rng.next_u64());
    let contract2 = ContractAddress::from(rng.next_u64());
    let class_hash = ClassHash(rng.next_u64().into());
    let compiled_class_hash = CompiledClassHash(rng.next_u64().into());
    let deprecated_class_hash = ClassHash(rng.next_u64().into());
    ThinStateDiff {
        deployed_contracts: indexmap! {
            contract0 => class_hash, contract1 => class_hash, contract2 => deprecated_class_hash
        },
        storage_diffs: indexmap! {
            contract0 => indexmap! {
                1u64.into() => Felt::ONE, 2u64.into() => Felt::TWO
            },
            contract1 => indexmap! {
                3u64.into() => Felt::TWO, 4u64.into() => Felt::ONE
            },
        },
        declared_classes: indexmap! { class_hash => compiled_class_hash },
        deprecated_declared_classes: vec![deprecated_class_hash],
        nonces: indexmap! {
            contract0 => Nonce(Felt::ONE), contract2 => Nonce(Felt::TWO)
        },
        replaced_classes: Default::default(),
    }
}
