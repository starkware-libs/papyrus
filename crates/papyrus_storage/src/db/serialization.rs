use crate::db::DbError;

pub(crate) trait StorageSerdeEx: StorageSerde {
    fn serialize(&self) -> Result<Vec<u8>, DbError>;

    fn deserialize(bytes: &mut impl std::io::Read) -> Option<Self>;
}

impl<T: StorageSerde> StorageSerdeEx for T {
    fn serialize(&self) -> Result<Vec<u8>, DbError> {
        let mut res: Vec<u8> = Vec::new();
        self.serialize_into(&mut res).map_err(|_| DbError::Serialization)?;
        Ok(res)
    }

    fn deserialize(bytes: &mut impl std::io::Read) -> Option<Self> {
        let res = Self::deserialize_from(bytes)?;
        let mut buf = [0u8, 1];
        // Make sure we are at EOF.
        if bytes.read(&mut buf[..]).ok()? != 0 {
            return None;
        }
        Some(res)
    }
}

pub trait StorageSerde: Sized {
    fn serialize_into(&self, res: &mut impl std::io::Write) -> Result<(), StorageSerdeError>;

    fn deserialize_from(bytes: &mut impl std::io::Read) -> Option<Self>;
}

#[derive(thiserror::Error, Debug)]
pub enum StorageSerdeError {
    #[error(transparent)]
    Bincode(#[from] bincode::Error),
    #[error(transparent)]
    Io(#[from] std::io::Error),
    #[error(transparent)]
    Serde(#[from] serde_json::Error),
}
