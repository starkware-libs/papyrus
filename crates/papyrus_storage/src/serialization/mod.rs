pub(crate) mod serializers;
#[cfg(test)]
pub(crate) mod serializers_test;
