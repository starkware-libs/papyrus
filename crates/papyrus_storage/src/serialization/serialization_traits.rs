//! Traits for serializing and deserializing objects for storing in a Papyrus storage.

use std::fmt::Debug;
use std::io::Write;
use std::marker::PhantomData;

/// Trait for serializing and deserializing values.
pub(crate) trait StorageSerdeEx: StorageSerde {
    fn serialize(&self) -> Result<Vec<u8>, StorageSerdeError>;

    fn deserialize(bytes: &mut impl std::io::Read) -> Option<Self>;
}

impl<T: StorageSerde> StorageSerdeEx for T {
    fn serialize(&self) -> Result<Vec<u8>, StorageSerdeError> {
        let mut res: Vec<u8> = Vec::new();
        self.serialize_into(&mut res)?;
        Ok(res)
    }

    fn deserialize(bytes: &mut impl std::io::Read) -> Option<Self> {
        let res = Self::deserialize_from(bytes)?;
        let mut buf = [0u8, 1];
        // Make sure we are at EOF.
        if bytes.read(&mut buf[..]).ok()? != 0 {
            return None;
        }
        Some(res)
    }
}

/// Trait for deserializing and serializing values into buffers.
pub trait StorageSerde: Sized {
    /// Serializes the object into the buffer.
    fn serialize_into(&self, res: &mut impl std::io::Write) -> Result<(), StorageSerdeError>;

    /// Deserializes the object from the buffer.
    fn deserialize_from(bytes: &mut impl std::io::Read) -> Option<Self>;
}

/// Trait that enforces a database key to implement `StorageSerdeEx`, `Ord` and `Clone`.
pub(crate) trait Key: StorageSerdeEx + Ord + Clone {}
impl<T> Key for T where T: StorageSerdeEx + Ord + Clone {}

/// Trait for serializing and deserializing values from the database.
pub(crate) trait ValueSerde {
    type Value: StorageSerde + Debug;

    fn serialize(obj: &Self::Value) -> Result<Vec<u8>, StorageSerdeError>;
    fn deserialize(bytes: &mut impl std::io::Read) -> Option<Self::Value>;
}

#[derive(Clone, Debug)]
/// A generic wrapper for values that do not have a version.
pub(crate) struct NoVersionValueWrapper<T: StorageSerde> {
    _value_type: PhantomData<T>,
}

impl<T: StorageSerde + Debug> ValueSerde for NoVersionValueWrapper<T> {
    type Value = T;

    fn serialize(obj: &Self::Value) -> Result<Vec<u8>, StorageSerdeError> {
        StorageSerdeEx::serialize(obj)
    }

    fn deserialize(bytes: &mut impl std::io::Read) -> Option<Self::Value> {
        StorageSerdeEx::deserialize(bytes)
    }
}

#[derive(Clone, Debug)]
/// A generic wrapper for values with version zero. These values are serialized with a leading byte
/// that is set to zero.
pub(crate) struct VersionZeroWrapper<T: StorageSerde> {
    _value_type: PhantomData<T>,
}

const VERSION_ZERO: u8 = 0;

impl<T: StorageSerde + Debug> ValueSerde for VersionZeroWrapper<T> {
    type Value = T;

    fn serialize(obj: &Self::Value) -> Result<Vec<u8>, StorageSerdeError> {
        let mut res = Vec::new();
        res.write_all(&[VERSION_ZERO]).expect("Failed to write version");
        obj.serialize_into(&mut res)?;
        Ok(res)
    }

    fn deserialize(bytes: &mut impl std::io::Read) -> Option<Self::Value> {
        let mut version = [0u8; 1];
        bytes.read_exact(&mut version[..]).ok()?;
        if version[0] != VERSION_ZERO {
            return None;
        }
        let res = Self::Value::deserialize_from(bytes)?;

        let mut buf = [0u8, 1];
        // Make sure we are at EOF.
        if bytes.read(&mut buf[..]).ok()? != 0 {
            return None;
        }
        Some(res)
    }
}

/// Error type for serialization and deserialization.
#[derive(thiserror::Error, Debug)]
pub enum StorageSerdeError {
    /// An error occurred during IO operations.
    #[error(transparent)]
    Io(#[from] std::io::Error),
    /// An error occurred during serialization or deserialization.
    #[error(transparent)]
    Serde(#[from] serde_json::Error),
}
