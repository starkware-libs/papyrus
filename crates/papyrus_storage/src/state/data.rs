use indexmap::IndexMap;
use serde::{Deserialize, Serialize};
use starknet_api::block::BlockNumber;
use starknet_api::core::{ClassHash, ContractAddress, Nonce};
use starknet_api::hash::StarkFelt;
use starknet_api::state::{ContractClass, StateDiff, StorageKey};

/// Data structs that are serialized into the database.

// Invariant: Addresses are strictly increasing.
// The invariant is enforced as [`ThinStateDiff`] is created only from [`starknet_api`][`StateDiff`]
// where the addresses are strictly increasing.
#[derive(Debug, Clone, Eq, PartialEq, Deserialize, Serialize)]
pub struct ThinStateDiff {
    pub deployed_contracts: IndexMap<ContractAddress, ClassHash>,
    pub storage_diffs: IndexMap<ContractAddress, IndexMap<StorageKey, StarkFelt>>,
    pub declared_contract_hashes: Vec<ClassHash>,
    pub nonces: IndexMap<ContractAddress, Nonce>,
}

impl ThinStateDiff {
    // Returns also the declared classes without cloning them.
    pub(crate) fn from_state_diff(diff: StateDiff) -> (Self, IndexMap<ClassHash, ContractClass>) {
        (
            Self {
                deployed_contracts: diff.deployed_contracts,
                storage_diffs: diff.storage_diffs,
                declared_contract_hashes: diff.declared_classes.keys().copied().collect(),
                nonces: diff.nonces,
            },
            diff.declared_classes,
        )
    }
}

impl From<StateDiff> for ThinStateDiff {
    fn from(diff: StateDiff) -> Self {
        Self::from_state_diff(diff).0
    }
}

#[derive(Debug, Default, Clone, Eq, PartialEq, Hash, Deserialize, Serialize, PartialOrd, Ord)]
pub(crate) struct IndexedDeployedContract {
    pub block_number: BlockNumber,
    pub class_hash: ClassHash,
}

#[derive(Debug, Default, Clone, Eq, PartialEq, Deserialize, Serialize)]
pub(crate) struct IndexedDeclaredContract {
    pub block_number: BlockNumber,
    pub contract_class: ContractClass,
}
