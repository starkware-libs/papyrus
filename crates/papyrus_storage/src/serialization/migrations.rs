use tracing::error;

use crate::db::serialization::{Migratable, StorageSerde, StorageSerdeError};
use crate::header::{StorageBlockHeader, StorageBlockHeaderV0};

impl Migratable for StorageBlockHeader {
    fn try_from_older_version(
        bytes: &mut impl std::io::Read,
        older_version: u8,
    ) -> Result<Self, StorageSerdeError> {
        if older_version != 0 {
            error!(
                "Unable to migrate stored header from version {} to current version.",
                older_version
            );
            return Err(StorageSerdeError::Migration);
        }
        let v0_header =
            StorageBlockHeaderV0::deserialize_from(bytes).ok_or(StorageSerdeError::Migration)?;
        Ok(v0_header.into())
    }
}
