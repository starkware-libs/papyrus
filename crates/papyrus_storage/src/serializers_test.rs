use std::fmt::Debug;

use starknet_api::core::ContractAddress;
use starknet_api::hash::StarkHash;
use starknet_api::state::StorageKey;
use test_utils::{get_rng, GetTestInstance};

use crate::db::serialization::StorageSerde;

pub trait StorageSerdeTest: StorageSerde {
    fn storage_serde_test();
}

// Implements the [`storage_serde_test`] function for every type that
// implements the [`StorageSerde`] and [`GetTestInstance`] traits.
impl<T: StorageSerde + GetTestInstance + Eq + Debug> StorageSerdeTest for T {
    fn storage_serde_test() {
        let mut rng = get_rng();
        let item = T::get_test_instance(&mut rng);
        let mut serialized: Vec<u8> = Vec::new();
        item.serialize_into(&mut serialized).unwrap();
        let bytes = serialized.into_boxed_slice();
        let deserialized = T::deserialize_from(&mut bytes.as_ref());
        assert_eq!(item, deserialized.unwrap());
    }
}

// Tests all types that implement the [`StorageSerde`] trait
// via the [`auto_storage_serde`] macro.
macro_rules! create_storage_serde_test {
    ($name:ident) => {
        paste::paste! {
            #[test]
            fn [<"storage_serde_test" _$name:snake>]() {
                $name::storage_serde_test()
            }
        }
    };
}
pub(crate) use create_storage_serde_test;

////////////////////////////////////////////////////////////////////////
// Implements the [`GetTestInstance`] trait for types not supported
// by the macro [`impl_get_test_instance`] and calls the [`create_test`]
// macro to create the tests for them.
////////////////////////////////////////////////////////////////////////
create_storage_serde_test!(StarkHash);
create_storage_serde_test!(ContractAddress);
create_storage_serde_test!(StorageKey);
