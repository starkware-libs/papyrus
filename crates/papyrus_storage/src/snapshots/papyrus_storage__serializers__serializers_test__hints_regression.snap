---
source: crates/papyrus_storage/src/serializers_test.rs
expression: casm.hints
---
- - 0
  - - TestLessThanOrEqual:
        lhs:
          Immediate: "0x1d10"
        rhs:
          Deref:
            register: FP
            offset: -6
        dst:
          register: AP
          offset: 0
- - 19
  - - AllocSegment:
        dst:
          register: AP
          offset: 0
- - 38
  - - TestLessThanOrEqual:
        lhs:
          Immediate: "0x0"
        rhs:
          Deref:
            register: AP
            offset: -8
        dst:
          register: AP
          offset: 0
- - 58
  - - AllocSegment:
        dst:
          register: AP
          offset: 0
- - 81
  - - AllocSegment:
        dst:
          register: AP
          offset: 0
- - 96
  - - AllocSegment:
        dst:
          register: AP
          offset: 0
- - 111
  - - TestLessThanOrEqual:
        lhs:
          Immediate: "0x1d10"
        rhs:
          Deref:
            register: FP
            offset: -6
        dst:
          register: AP
          offset: 0
- - 130
  - - AllocSegment:
        dst:
          register: AP
          offset: 0
- - 149
  - - TestLessThanOrEqual:
        lhs:
          Immediate: "0x0"
        rhs:
          Deref:
            register: AP
            offset: -8
        dst:
          register: AP
          offset: 0
- - 169
  - - AllocSegment:
        dst:
          register: AP
          offset: 0
- - 192
  - - AllocSegment:
        dst:
          register: AP
          offset: 0
- - 207
  - - AllocSegment:
        dst:
          register: AP
          offset: 0
- - 222
  - - TestLessThanOrEqual:
        lhs:
          Immediate: "0x2486"
        rhs:
          Deref:
            register: FP
            offset: -6
        dst:
          register: AP
          offset: 0
- - 241
  - - AllocSegment:
        dst:
          register: AP
          offset: 0
- - 260
  - - TestLessThanOrEqual:
        lhs:
          Immediate: "0x0"
        rhs:
          Deref:
            register: AP
            offset: -8
        dst:
          register: AP
          offset: 0
- - 280
  - - AllocSegment:
        dst:
          register: AP
          offset: 0
- - 300
  - - AllocSegment:
        dst:
          register: AP
          offset: 0
- - 315
  - - AllocSegment:
        dst:
          register: AP
          offset: 0
- - 330
  - - TestLessThanOrEqual:
        lhs:
          Immediate: "0x6054"
        rhs:
          Deref:
            register: FP
            offset: -6
        dst:
          register: AP
          offset: 0
- - 349
  - - AllocSegment:
        dst:
          register: AP
          offset: 0
- - 368
  - - TestLessThanOrEqual:
        lhs:
          Immediate: "0x0"
        rhs:
          Deref:
            register: AP
            offset: -8
        dst:
          register: AP
          offset: 0
- - 388
  - - AllocSegment:
        dst:
          register: AP
          offset: 0
- - 412
  - - AllocSegment:
        dst:
          register: AP
          offset: 0
- - 427
  - - AllocSegment:
        dst:
          register: AP
          offset: 0
- - 442
  - - TestLessThanOrEqual:
        lhs:
          Immediate: "0x7cec"
        rhs:
          Deref:
            register: FP
            offset: -6
        dst:
          register: AP
          offset: 0
- - 467
  - - AllocSegment:
        dst:
          register: AP
          offset: 0
- - 490
  - - TestLessThanOrEqual:
        lhs:
          Deref:
            register: AP
            offset: -1
        rhs:
          Deref:
            register: AP
            offset: -40
        dst:
          register: AP
          offset: 0
- - 510
  - - AllocSegment:
        dst:
          register: AP
          offset: 0
- - 536
  - - AllocSegment:
        dst:
          register: AP
          offset: 0
- - 552
  - - AllocSegment:
        dst:
          register: AP
          offset: 0
- - 567
  - - AllocSegment:
        dst:
          register: AP
          offset: 0
- - 583
  - - TestLessThanOrEqual:
        lhs:
          Immediate: "0x8fb6"
        rhs:
          Deref:
            register: FP
            offset: -6
        dst:
          register: AP
          offset: 0
- - 615
  - - AllocSegment:
        dst:
          register: AP
          offset: 0
- - 640
  - - TestLessThanOrEqual:
        lhs:
          Deref:
            register: AP
            offset: -1
        rhs:
          Deref:
            register: AP
            offset: -72
        dst:
          register: AP
          offset: 0
- - 661
  - - AllocSegment:
        dst:
          register: AP
          offset: 0
- - 687
  - - AllocSegment:
        dst:
          register: AP
          offset: 0
- - 703
  - - AllocSegment:
        dst:
          register: AP
          offset: 0
- - 718
  - - AllocSegment:
        dst:
          register: AP
          offset: 0
- - 733
  - - AllocSegment:
        dst:
          register: AP
          offset: 0
- - 749
  - - TestLessThanOrEqual:
        lhs:
          Immediate: "0x2bf2a"
        rhs:
          Deref:
            register: FP
            offset: -6
        dst:
          register: AP
          offset: 0
- - 781
  - - AllocSegment:
        dst:
          register: AP
          offset: 0
- - 806
  - - TestLessThanOrEqual:
        lhs:
          Deref:
            register: AP
            offset: -1
        rhs:
          Deref:
            register: AP
            offset: -93
        dst:
          register: AP
          offset: 0
- - 828
  - - AllocSegment:
        dst:
          register: AP
          offset: 0
- - 848
  - - AllocSegment:
        dst:
          register: AP
          offset: 0
- - 864
  - - AllocSegment:
        dst:
          register: AP
          offset: 0
- - 879
  - - AllocSegment:
        dst:
          register: AP
          offset: 0
- - 894
  - - AllocSegment:
        dst:
          register: AP
          offset: 0
- - 910
  - - TestLessThanOrEqual:
        lhs:
          Immediate: "0x448d6"
        rhs:
          Deref:
            register: FP
            offset: -6
        dst:
          register: AP
          offset: 0
- - 949
  - - AllocSegment:
        dst:
          register: AP
          offset: 0
- - 974
  - - TestLessThanOrEqual:
        lhs:
          Deref:
            register: AP
            offset: -1
        rhs:
          Deref:
            register: AP
            offset: -124
        dst:
          register: AP
          offset: 0
- - 997
  - - AllocSegment:
        dst:
          register: AP
          offset: 0
- - 1017
  - - AllocSegment:
        dst:
          register: AP
          offset: 0
- - 1033
  - - AllocSegment:
        dst:
          register: AP
          offset: 0
- - 1048
  - - AllocSegment:
        dst:
          register: AP
          offset: 0
- - 1063
  - - AllocSegment:
        dst:
          register: AP
          offset: 0
- - 1078
  - - AllocSegment:
        dst:
          register: AP
          offset: 0
- - 1094
  - - TestLessThanOrEqual:
        lhs:
          Immediate: "0x11ea4"
        rhs:
          Deref:
            register: FP
            offset: -6
        dst:
          register: AP
          offset: 0
- - 1126
  - - AllocSegment:
        dst:
          register: AP
          offset: 0
- - 1151
  - - TestLessThanOrEqual:
        lhs:
          Deref:
            register: AP
            offset: -1
        rhs:
          Deref:
            register: AP
            offset: -93
        dst:
          register: AP
          offset: 0
- - 1173
  - - AllocSegment:
        dst:
          register: AP
          offset: 0
- - 1193
  - - AllocSegment:
        dst:
          register: AP
          offset: 0
- - 1209
  - - AllocSegment:
        dst:
          register: AP
          offset: 0
- - 1224
  - - AllocSegment:
        dst:
          register: AP
          offset: 0
- - 1239
  - - AllocSegment:
        dst:
          register: AP
          offset: 0
- - 1255
  - - TestLessThanOrEqual:
        lhs:
          Immediate: "0x1b9f4"
        rhs:
          Deref:
            register: FP
            offset: -6
        dst:
          register: AP
          offset: 0
- - 1287
  - - AllocSegment:
        dst:
          register: AP
          offset: 0
- - 1312
  - - TestLessThanOrEqual:
        lhs:
          Deref:
            register: AP
            offset: -1
        rhs:
          Deref:
            register: AP
            offset: -93
        dst:
          register: AP
          offset: 0
- - 1334
  - - AllocSegment:
        dst:
          register: AP
          offset: 0
- - 1354
  - - AllocSegment:
        dst:
          register: AP
          offset: 0
- - 1370
  - - AllocSegment:
        dst:
          register: AP
          offset: 0
- - 1385
  - - AllocSegment:
        dst:
          register: AP
          offset: 0
- - 1400
  - - AllocSegment:
        dst:
          register: AP
          offset: 0
- - 1416
  - - TestLessThanOrEqual:
        lhs:
          Immediate: "0x1b9f4"
        rhs:
          Deref:
            register: FP
            offset: -6
        dst:
          register: AP
          offset: 0
- - 1448
  - - AllocSegment:
        dst:
          register: AP
          offset: 0
- - 1473
  - - TestLessThanOrEqual:
        lhs:
          Deref:
            register: AP
            offset: -1
        rhs:
          Deref:
            register: AP
            offset: -93
        dst:
          register: AP
          offset: 0
- - 1495
  - - AllocSegment:
        dst:
          register: AP
          offset: 0
- - 1515
  - - AllocSegment:
        dst:
          register: AP
          offset: 0
- - 1531
  - - AllocSegment:
        dst:
          register: AP
          offset: 0
- - 1546
  - - AllocSegment:
        dst:
          register: AP
          offset: 0
- - 1561
  - - AllocSegment:
        dst:
          register: AP
          offset: 0
- - 1577
  - - TestLessThanOrEqual:
        lhs:
          Immediate: "0x1fa36"
        rhs:
          Deref:
            register: FP
            offset: -6
        dst:
          register: AP
          offset: 0
- - 1649
  - - AllocSegment:
        dst:
          register: AP
          offset: 0
- - 1672
  - - TestLessThanOrEqual:
        lhs:
          Deref:
            register: AP
            offset: -1
        rhs:
          Deref:
            register: AP
            offset: -142
        dst:
          register: AP
          offset: 0
- - 1697
  - - AllocSegment:
        dst:
          register: AP
          offset: 0
- - 1717
  - - AllocSegment:
        dst:
          register: AP
          offset: 0
- - 1733
  - - AllocSegment:
        dst:
          register: AP
          offset: 0
- - 1748
  - - AllocSegment:
        dst:
          register: AP
          offset: 0
- - 1769
  - - AllocSegment:
        dst:
          register: AP
          offset: 0
- - 1784
  - - AllocSegment:
        dst:
          register: AP
          offset: 0
- - 1799
  - - AllocSegment:
        dst:
          register: AP
          offset: 0
- - 1814
  - - AllocSegment:
        dst:
          register: AP
          offset: 0
- - 1936
  - - TestLessThan:
        lhs:
          Deref:
            register: AP
            offset: -1
        rhs:
          Immediate: "0x800000000000000000000000000000000000000000000000000000000000000"
        dst:
          register: AP
          offset: 4
- - 1940
  - - LinearSplit:
        value:
          Deref:
            register: AP
            offset: 3
        scalar:
          Immediate: "0x110000000000000000"
        max_x:
          Immediate: "0xffffffffffffffffffffffffffffffff"
        x:
          register: AP
          offset: -2
        y:
          register: AP
          offset: -1
- - 1950
  - - LinearSplit:
        value:
          Deref:
            register: AP
            offset: -2
        scalar:
          Immediate: "0x8000000000000000000000000000000"
        max_x:
          Immediate: "0xffffffffffffffffffffffffffffffff"
        x:
          register: AP
          offset: -1
        y:
          register: AP
          offset: 0
- - 2530
  - - TestLessThan:
        lhs:
          Deref:
            register: FP
            offset: -3
        rhs:
          Immediate: "0x100"
        dst:
          register: AP
          offset: 0
- - 2534
  - - LinearSplit:
        value:
          Deref:
            register: AP
            offset: -1
        scalar:
          Immediate: "0x8000000000000110000000000000000"
        max_x:
          Immediate: "0xfffffffffffffffffffffffffffffffe"
        x:
          register: AP
          offset: 0
        y:
          register: AP
          offset: 1
- - 2606
  - - AllocSegment:
        dst:
          register: AP
          offset: 0
- - 2738
  - - SystemCall:
        system:
          Deref:
            register: FP
            offset: -3
- - 2787
  - - SystemCall:
        system:
          Deref:
            register: FP
            offset: -3
- - 2836
  - - SystemCall:
        system:
          Deref:
            register: FP
            offset: -3
- - 2860
  - - AllocSegment:
        dst:
          register: AP
          offset: 0
- - 3052
  - - TestLessThan:
        lhs:
          Deref:
            register: FP
            offset: -3
        rhs:
          Immediate: "0x100000000000000000000000000000000"
        dst:
          register: AP
          offset: 0
- - 3054
  - - DivMod:
        lhs:
          Deref:
            register: FP
            offset: -3
        rhs:
          Immediate: "0x100000000000000000000000000000000"
        quotient:
          register: AP
          offset: 3
        remainder:
          register: AP
          offset: 4
- - 3129
  - - AllocSegment:
        dst:
          register: AP
          offset: 0
- - 3161
  - - AllocSegment:
        dst:
          register: AP
          offset: 0
- - 3463
  - - AllocSegment:
        dst:
          register: AP
          offset: 0
- - 3548
  - - AllocSegment:
        dst:
          register: AP
          offset: 0
- - 3577
  - - AllocSegment:
        dst:
          register: AP
          offset: 0
- - 3600
  - - SystemCall:
        system:
          Deref:
            register: FP
            offset: -4
- - 3652
  - - SystemCall:
        system:
          Deref:
            register: FP
            offset: -4
- - 3704
  - - SystemCall:
        system:
          Deref:
            register: FP
            offset: -4
- - 3820
  - - AllocSegment:
        dst:
          register: AP
          offset: 0
- - 3822
  - - AllocSegment:
        dst:
          register: AP
          offset: 0
- - 3847
  - - SystemCall:
        system:
          Deref:
            register: FP
            offset: -8
- - 3920
  - - SystemCall:
        system:
          Deref:
            register: FP
            offset: -5
- - 3940
  - - SystemCall:
        system:
          Deref:
            register: AP
            offset: -3
- - 3964
  - - AllocSegment:
        dst:
          register: AP
          offset: 0
- - 3995
  - - AllocSegment:
        dst:
          register: AP
          offset: 0
- - 4041
  - - TestLessThan:
        lhs:
          Deref:
            register: AP
            offset: -1
        rhs:
          Immediate: "0x7ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff00"
        dst:
          register: AP
          offset: 5
- - 4045
  - - LinearSplit:
        value:
          Deref:
            register: AP
            offset: 4
        scalar:
          Immediate: "0x110000000000000000"
        max_x:
          Immediate: "0xffffffffffffffffffffffffffffffff"
        x:
          register: AP
          offset: -2
        y:
          register: AP
          offset: -1
- - 4056
  - - LinearSplit:
        value:
          Deref:
            register: AP
            offset: 4
        scalar:
          Immediate: "0x8000000000000000000000000000000"
        max_x:
          Immediate: "0xfffffffffffffffffffffffffffffffe"
        x:
          register: AP
          offset: -2
        y:
          register: AP
          offset: -1
- - 4084
  - - TestLessThan:
        lhs:
          Deref:
            register: AP
            offset: -1
        rhs:
          Immediate: "0x7ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff00"
        dst:
          register: AP
          offset: 5
- - 4088
  - - LinearSplit:
        value:
          Deref:
            register: AP
            offset: 4
        scalar:
          Immediate: "0x110000000000000000"
        max_x:
          Immediate: "0xffffffffffffffffffffffffffffffff"
        x:
          register: AP
          offset: -2
        y:
          register: AP
          offset: -1
- - 4099
  - - LinearSplit:
        value:
          Deref:
            register: AP
            offset: 4
        scalar:
          Immediate: "0x8000000000000000000000000000000"
        max_x:
          Immediate: "0xfffffffffffffffffffffffffffffffe"
        x:
          register: AP
          offset: -2
        y:
          register: AP
          offset: -1
- - 4123
  - - SystemCall:
        system:
          Deref:
            register: FP
            offset: -3
- - 4168
  - - AllocSegment:
        dst:
          register: AP
          offset: 0
- - 4170
  - - AllocSegment:
        dst:
          register: AP
          offset: 0
- - 4195
  - - SystemCall:
        system:
          Deref:
            register: FP
            offset: -7
- - 4281
  - - AllocSegment:
        dst:
          register: AP
          offset: 0
- - 4283
  - - AllocSegment:
        dst:
          register: AP
          offset: 0
- - 4308
  - - SystemCall:
        system:
          Deref:
            register: FP
            offset: -7
- - 4422
  - - SystemCall:
        system:
          Deref:
            register: FP
            offset: -7
- - 4435
  - - SystemCall:
        system:
          BinOp:
            op: Add
            a:
              register: FP
              offset: -7
            b:
              Immediate: "0x7"
- - 4541
  - - TestLessThan:
        lhs:
          Deref:
            register: AP
            offset: 0
        rhs:
          Immediate: "0x100000000000000000000000000000000"
        dst:
          register: AP
          offset: -1
- - 4564
  - - TestLessThan:
        lhs:
          Deref:
            register: AP
            offset: 0
        rhs:
          Immediate: "0x100000000000000000000000000000000"
        dst:
          register: AP
          offset: -1
- - 4584
  - - TestLessThan:
        lhs:
          Deref:
            register: AP
            offset: 0
        rhs:
          Immediate: "0x100000000000000000000000000000000"
        dst:
          register: AP
          offset: -1
- - 4609
  - - TestLessThan:
        lhs:
          Deref:
            register: AP
            offset: 0
        rhs:
          Immediate: "0x100000000000000000000000000000000"
        dst:
          register: AP
          offset: -1
- - 4632
  - - TestLessThan:
        lhs:
          Deref:
            register: AP
            offset: 0
        rhs:
          Immediate: "0x100000000000000000000000000000000"
        dst:
          register: AP
          offset: -1
- - 4652
  - - TestLessThan:
        lhs:
          Deref:
            register: AP
            offset: 0
        rhs:
          Immediate: "0x100000000000000000000000000000000"
        dst:
          register: AP
          offset: -1

